### PR TITLE
fix(theme-classic): avoid rendering empty search box container

### PR DESCRIPTION
**Pull request by drizzlesconsin**
## Pre-flight checklist

- [x] I have read the [Contributing Guidelines on pull requests](https://github.com/facebook/docusaurus/blob/main/CONTRIBUTING.md#pull-requests).
- [x] **If this is a code change**: I have written unit tests and/or added dogfooding pages to fully verify the new behavior.
- [ ] **If this is a new API or substantial change**: the PR has an accompanying issue (closes #0000) and the maintainers have approved on my working plan.

## Motivation

image

https://stackblitz.com/github/facebook/docusaurus/tree/starter

The search box should not be displayed when the search function is not enabled.

### Test links

Deploy preview: https://deploy-preview-_____--docusaurus-2.netlify.app/

### CODE REVIEW

File: packages/docusaurus-theme-classic/src/theme/Navbar/Search/index.tsx [hunk 1]
```diff
@@ -15,5 +15,6 @@ export default function NavbarSearch({
   children,
   className,
 }: Props): JSX.Element {
+  if (!children) return <></>;
```
Comment (Josh-Cena) [inline packages/docusaurus-theme-classic/src/theme/Navbar/Search/index.tsx:18]:
```suggestion
  if (!children) return null;
```

File: packages/docusaurus-theme-classic/src/theme/Navbar/Search/index.tsx [hunk 2]
```diff
@@ -15,5 +15,8 @@ export default function NavbarSearch({
   children,
   className,
 }: Props): JSX.Element {
```
Comment (Josh-Cena) [inline packages/docusaurus-theme-classic/src/theme/Navbar/Search/index.tsx:17]:
```suggestion
}: Props): JSX.Element | null {
```

**Review by Josh-Cena — APPROVED** [review]
Thank you!

Just pending the CLA

**Review by Josh-Cena — CHANGES_REQUESTED** [review]
Actually, this doesn't work. If you check locally, the empty container is still rendered. This is because the `<SearchBar />` element is always there; it just sometimes returns `null` (which is not the same as a literal `null` children). We need a better solution.

**Comment by drizzlesconsin** [comment]
https://github.com/facebook/docusaurus/blob/main/packages/docusaurus-theme-search-algolia/src/theme/SearchBar/index.tsx#L268-L270

That can only be determined by the `algolia` configuration.

File: packages/docusaurus-theme-classic/src/theme/Navbar/Search/index.tsx [hunk 3]
```diff
@@ -7,13 +7,18 @@
...
 
 import styles from './styles.module.css';
 
 export default function NavbarSearch({
   children,
   className,
-}: Props): JSX.Element {
+}: Props): JSX.Element | null {
+  const {siteConfig} = useDocusaurusContext();
+  if (!siteConfig.themeConfig?.algolia) {
```
Comment (slorber) [inline packages/docusaurus-theme-classic/src/theme/Navbar/Search/index.tsx:20]: this will not work for non-official search plugins

**Review by slorber — CHANGES_REQUESTED** [review]

**Comment by slorber** [comment]
The problem is that we are trying to add an implicit search bar by default (ie not configured through `themeConfig.navbar.items`). When user does not want search, it does not make sense.

What I suggest is that in 3.0 we can do a breaking change:
- stop automatically adding an implicit search-bar, force user to be explicit and always provide it as a navbar item
- warn if user has no search navbar item but there's a an algolia config provided, eventually try to restore a retrocompatible behavior?
- make sure it works nicely on mobile too!

Another possible solution: using CSS `.myContainer:has(> *)` but it's not supported enough unfortunately

In the future, we should allow search plugin authors to register their own Navbar item components. They would be able to create a `type: "my-local-search"` instead of overwriting `@theme/SearchBar`

File: packages/docusaurus-theme-classic/src/theme/Navbar/Search/index.tsx (hunk 3 above)
Comment (drizzlesconsin) [reply packages/docusaurus-theme-classic/src/theme/Navbar/Search/index.tsx:20]: em...Let's see if there's another way.

**Comment by drizzlesconsin** [comment]
Is it possible to use this method here?
```ts
children.type() === null
```
I read:
[React component children detect if empty / null before render](https://stackoverflow.com/a/64396494/7324216)
[Is there any way for Parent to know if Child rendered null?](https://twitter.com/markdalgleish/status/1471007400647790595)

**Comment by slorber** [comment]
There's no good solution to this problem in React. I don't want to introduce hacky workarounds in our theme when there's a clear elegant solution: be explicit and avoid rendering this parent search component in the first place

**Comment by drizzlesconsin** [comment]
Looks like it can only be hack with CSS.

**Comment by lex111** [comment]
What about using the named Noop component and then before rendering SearchBar, checking the value of `SearchBar.name !== 'Noop'`? Maybe is a lesser of an evil?

**Comment by slorber** [comment]
will be fixed in https://github.com/facebook/docusaurus/pull/9385

**Pull request closed without merging**